rqt_controller_manager/rqt_controller_manager/controller_manager.py

```diff
@@ -271,42 +295,207 @@ def _on_ctrl_info(self, index):
+    def _on_ctrl_header_menu(self, pos):
+        ctrl_header = self._widget.ctrl_table_view.horizontalHeader()
 
         # Show context menu
-        menu = QMenu(self._widget.table_view)
+        menu = QMenu(self._widget.ctrl_table_view)
         action_toggle_auto_resize = menu.addAction("Toggle Auto-Resize")
-        action = menu.exec_(header.mapToGlobal(pos))
+        action = menu.exec_(ctrl_header.mapToGlobal(pos))
 
         # Evaluate user action
         if action is action_toggle_auto_resize:
-            if header.resizeMode(0) == QHeaderView.ResizeToContents:
-                header.setSectionResizeMode(QHeaderView.Interactive)
+            if ctrl_header.resizeMode(0) == QHeaderView.ResizeToContents:
+                ctrl_header.setSectionResizeMode(QHeaderView.Interactive)
             else:
-                header.setSectionResizeMode(QHeaderView.ResizeToContents)
+                ctrl_header.setSectionResizeMode(QHeaderView.ResizeToContents)
+
+    def _update_hw_components(self):
+        if not self._cm_name:
+            return
+
+        # Find hw_components associated to the selected controller manager
+        hw_components = self._list_hw_components()
+
+        # Update controller display, if necessary
+        if self._hw_components != hw_components:
+            self._hw_components = hw_components
+            self._show_hw_components()  # NOTE: Model is recomputed from scratch
+
+    def _list_hw_components(self):
+        """
+        List the hw_components associated to a controller manager node.
+
+        @return List of hw_components associated to a controller manager
+        node. Contains both stopped/running hw_components, as returned by
+        the C{list_hardware_components} service
+        @rtype [str]
+        """
+        # Add loaded hw_components first
+        try:
+            hw_components = list_hardware_components(
+                self._node, self._cm_name, 2.0 / self._cm_update_freq
+            ).component
+            return hw_components
+        except RuntimeError as e:
+            print(e)
+            return []
+
+    def _show_hw_components(self):
+        hw_table_view = self._widget.hw_table_view
+        self._hw_table_model = HwComponentTable(self._hw_components, self._icons)
+        hw_table_view.setModel(self._hw_table_model)
+
+    def _on_hw_menu(self, pos):
+        # Get data of selected controller
+        row = self._widget.hw_table_view.rowAt(pos.y())
+        if row < 0:
+            return  # Cursor is not under a valid item
+
+        hw_component = self._hw_components[row]
+
+        # Show context menu
+        menu = QMenu(self._widget.hw_table_view)
+        if hw_component.state.label == "active":
+            action_deactivate = menu.addAction(self._icons["inactive"], "Deactivate")
+            action_shutdown = menu.addAction(self._icons["finalized"], "Shutdown")
+        elif hw_component.state.label == "inactive":
```

Comment:
you are right. I saw now that I used the wrong labels, instead of shutdown it should be cleanup, because we land in [unconfigured state instead of finalized](https://design.ros2.org/articles/node_lifecycle.html) -> I can remove the finalized of L377

Reply:
Awesome

Reply:
fixed, what do you think?

Reply:
I like it. Deactivate and cleanup :)